---
title: "RNA-seq analysis - Differential Expression"
author: "Jorge Peña Díaz"
date: '`r Sys.Date()`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

-----------------------------------------

Analysis used to compare the gene expression profile of the ∆*croI* and wild-type (WT) *C. rodentium* DB1000 strains as published in:

> Peña‑Díaz, J., Woodward, S.E., Creus‑Cuadros, A., Serapio‑Palacios, A., Deng, W., and Finlay, B.B. (2023). Quorum Sensing Modulates
Bacterial Virulence and Colonization Dynamics During an Enteric Infection. bioRxiv, [10.1101/2023.03.14.532656](https://www.biorxiv.org/content/10.1101/2023.03.14.532656v1).


-----------------------------------------

### 1. Setup

Load the required packages for the analysis.

```{r eval = TRUE, message=FALSE, warning=FALSE}
library(tidyverse)
library(DESeq2)
library(pheatmap)
library(ggsci)
library(IHW)
library(ggrepel)
library(WGCNA)
#library(bbcRNA)
```

Load the required custom functions.

**Note:** Since the installation of the `bbcRNA` package is not functioning, we'll import the required `star_to_mat` as a custom function.

```{r eval = TRUE, message=FALSE}
source("scripts/star_to_mat.R")
source("scripts/make_bold_names.R")
```

### 2. Import the raw data and metadata

```{r Raw data import, warning=FALSE}
# Import raw data
data <- star_to_mat(dir = "data/ReadsPerGene/", rgx = "^[^-]+", column = 1) 
head(data)

# Import metadata
metadata <- read.csv("data/metadata.csv", row.names = "name") %>%
  mutate(strain = factor(strain, levels = c("WT", "delta_croI")),
         replicate = factor(replicate))

head(metadata)
```

### 3. Pre-filtering and clean-up


```{r Pre-filtering}
# Check dimensions of the data
dim(data)

# Filter reads present in at least 3 samples with a count of 10 or higher
keep <- rowSums(data >= 10) >= 3
data <- data[keep,]

# Check dimensions again to see the effect of filtering
dim(data)


# Make data match the order of the metadata before importing into DESeq2

# Get order from the metadata
idx <- match(rownames(metadata), colnames(data))
data <- data[,idx]

# Check if order of the data matches the one from the metadata
all(rownames(metadata) == colnames(data))
```


### 3. DESeq2 

Create a DESeq2 object.

```{r DESeq2}
dds <- DESeqDataSetFromMatrix(countData =  data,
                              colData =  metadata,
                              design = ~ replicate + strain)
```

Perform count normalization to account for factors such as library depth, gene length or RNA composition.

```{r Count normalization}
dds <- estimateSizeFactors(dds)

dds_normalized_counts <- counts(dds, normalized = TRUE)
```

Perform a hierarchical heatmap.

```{r Heatmap}

# First it is necessary to log transform to improve visualization. 
vsd <- vst(dds, blind = TRUE) 

# Compute pairwise correlations and plot the heatmap
vsd %>%
  assay() %>%
  cor() %>%
  pheatmap(annotation = select(metadata, strain))
```

Perform clustering analyses.

```{r PCoA}

# Principal component analysis
pcaData <- plotPCA(vsd, intgroup = c("strain"), returnData=TRUE) 

# Extract variance in order to plot it on the axis legend
percentVar <- round(100 * attr(pcaData, "percentVar"))

# Plot
pcaData %>%
  mutate(strain = str_replace(strain, "delta_", "∆")) %>%
  mutate(strain = factor(strain, levels = c("WT", "∆croI"))) %>%
  mutate(replicate = str_sub(name, -1)) %>%
  ggplot(aes(x = PC1, y = PC2, fill = strain)) +
  geom_point(size = 3, colour = "black", shape = 21) +
  theme_bw() +
  scale_fill_npg() +
  labs(fill = "Strain") +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme(axis.title = element_text(size = rel(1.4)),
        axis.text = element_text(size = rel(1.4)),
        legend.title = element_text(size = rel(1.4)),
        legend.text = element_text(size = rel(1.2)))
```


Run DESeq2 analysis.

```{r Run DESeq2, message=FALSE}
dds <- DESeq(dds)
```


Test how well the data fits the model.

```{r Dispersion}
plotDispEsts(dds)
```


### 4. Import gene annotation for *C. rodentium* 

To create the annotation file, we combined data from the NCBI Reference Sequence Database ([RefSeq](https://www.ncbi.nlm.nih.gov/refseq/)) with information obtained from the Kyoto Encyclopedia of Genes and Genomes database ([KEGG](https://www.genome.jp/kegg/)).

```{r Annotation}
annotation <- read.csv("data/annotation_merged.csv")
glimpse(annotation)
```

### 5. Differential expression

We then performed differential expression with significance cutoff of 0.05 and using Independent Hypothesis Weighting (IHW) for p-value adjustment.

```{r Differential expression}

croI_res <- results(dds, 
                    filterFun=ihw,
                    contrast = c("strain", "delta_croI", "WT"),
                    alpha = 0.05)

summary(croI_res) # Summary stadistics
```

Perform log-fold chance shrinkage by using the 'apeglm' method

```{r LFC shrinkage, message=FALSE}
# Mean of the of the normalized counts compared to the log2 fold change for all genes that were tested
plotMA(croI_res, ylim = c(-8,8))

# To improve the estimates we are performing LFC shrinkage
croI_res_shrink <- lfcShrink(dds, coef = "strain_delta_croI_vs_WT",
                      res = croI_res, type = "apeglm")

# Shrunken estimates
plotMA(croI_res_shrink, ylim = c(-8,8))

# Save the results
croI_res_shrink_tidy <- croI_res_shrink %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  select(gene_id, lfc_MAP = log2FoldChange, lfcSE_MAP = lfcSE)
```

Final list of differentially regulated genes

```{r DEGs}
croI_res_sig <- subset(croI_res, padj < 0.05) %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) > log2(1.25)) %>%
  rownames_to_column("gene_id") %>%
  left_join(croI_res_shrink_tidy, by = "gene_id") %>%
  relocate(gene_id, baseMean, log2FoldChange, lfcSE, lfc_MAP, lfcSE_MAP) %>%
  left_join(annotation, by = "gene_id") %>%
  select(mapping = "Plasmid", gene_id, old_locus_tag, gene_name = "Gene", gene_name_synonym = "synonym", 
         product, product_kegg, baseMean, log2FoldChange, lfcSE, lfc_MAP, lfcSE_MAP,  stat, pvalue, padj,
         weight, gene_biotype, note) %>%
  arrange((padj))

glimpse(croI_res_sig)

# Export the result table
croI_res_sig %>%
  as.data.frame() %>%
  write.table("data/DEGs_dcroI.txt", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)
```


### 6. Visualization - Volcano plots 

First, we will gather the differential expression data for all genes. Then we'll merge it with the calculated shrunken LFC. Then the genes were classified as differentially regulated when the comparison between ∆*croI* and WT passed an adjusted *p*-value of < 0.05 and had a fold change equal or greater than ± 1.25. 

```{r Volcano plot}

croI_res_all <- croI_res %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  left_join(croI_res_shrink_tidy, by = "gene_id") %>%
  relocate(gene_id, baseMean, log2FoldChange, lfcSE, lfc_MAP, lfcSE_MAP) %>%
  left_join(annotation, by = "gene_id") %>%
  mutate(threshold = case_when(padj <= 0.05 & lfc_MAP >= log2(1.25) ~ "Upregulated",
                               padj <= 0.05 & lfc_MAP <= -log2(1.25) ~ "Downregulated",
                            padj > 0.05 | lfc_MAP < log2(1.25) & lfc_MAP > -log2(1.25) ~ "Not Significant")) %>%
  mutate(threshold = factor(threshold, levels = c("Upregulated", "Downregulated", "Not Significant"))) %>%
  arrange(padj) %>% 
  mutate(genelabels = "")

# The top 20 significantly differentially expressed genes ordered by ascending p-value were labelled
croI_res_all$genelabels[1:20] <- croI_res_all$Gene[1:20]
croI_res_all$genelabels[1] <- "putative lyase" 

# Confirm all the top 20 genes are differentially regulated
croI_res_all <- croI_res_all %>%
  mutate(genelabels = ifelse(str_detect(threshold, "Upregulated") | str_detect(threshold, "Downregulated"),  
                             genelabels, ""))

# Plot the volcano plot
ggplot(croI_res_all) +
  geom_point(aes(x = lfc_MAP, y = -log10(padj), fill = threshold),
             shape = 21, colour = "black", stroke = 0.0) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", alpha = 0.6) +
  geom_vline(xintercept = c(-log2(1.25), log2(1.25)),linetype = "dashed", alpha = 0.6) +
  geom_label_repel(aes(label = genelabels, x = log2FoldChange, y = -log10(padj)), max.overlaps = 18, size = 3.5, fontface = "italic") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  labs(fill = "Expression") +
  ggtitle("Differentially expressed genes", subtitle = "∆croI vs WT") +
  theme_bw() +
  scale_fill_manual(values = c("#FF3300", "#0D8CFF", "gray80")) +
  theme(plot.title = element_text(size = rel(1.9), hjust = 0.5),
        plot.subtitle = element_text(size = rel(1.35), hjust = 0.5),
        axis.title = element_text(size = rel(1.4)),
        axis.text = element_text(size = rel(1.4)),
        legend.title = element_text(size = rel(1.4)),
        legend.text = element_text(size = rel(1.2)))

# Export gene expression data for pathway enrichment analysis
croI_res_all %>%
  write.table("data/ExpressionData_dcroI.txt", sep="\t", col.names = TRUE, row.names = FALSE, quote = FALSE)
```


### 7. Visualization - Heatmap

Create a heatmap showing Z-score expression levels of genes associated with the type III secretion system (T3SS).

```{r Visualization - Heatmap, fig.dim = c(6, 8)}
# Import the clean annotation for the T3SS
T3S_operon <- read.csv("data/T3S_operon.csv")

# Clean metadata
labels_cols <- metadata %>%
  mutate(strain = str_replace(strain, "delta_croI", "∆croI")) %>%
  mutate(strain = factor(strain, levels = c("WT", "∆croI"))) %>%
  mutate(strain = paste0(strain, "_", replicate))

# Normalized counts for the T3SS
croI_norm_counts <- dds_normalized_counts %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  right_join(T3S_operon, by = "gene_id") %>%
  arrange(desc(order))

# Highlight which genes are differentially expressed
sig_norm_counts_croI_annotation <- croI_norm_counts %>%
  left_join(croI_res_all %>% select(gene_id, threshold), by = "gene_id") %>%
  mutate(gene_name = ifelse(threshold == "Upregulated" | threshold == "Downregulated",  paste0(gene_name,"*"), gene_name)) %>%
  select(gene_id, gene_name) %>%
  column_to_rownames("gene_name")

# Mark which genes to bold
bold_genes <- sig_norm_counts_croI_annotation %>%
  rownames_to_column() %>%
  filter(str_detect(rowname, "\\*")) %>%
  select(rowname) 

# Transform data to the needed format
croI_norm_counts <- croI_norm_counts %>%
  column_to_rownames("gene_id") %>%
  select(c(1:6))

# Create a colour pallete
heat_colors <- blueWhiteRed(200, gamma = 0.5)

# Plot the heatmap
pheatmap(croI_norm_counts,
         cellwidth = 35,
         color = heat_colors,
         cluster_rows = F,
         cluster_cols = F,
         show_rownames = T,
         fontsize = 11,
         labels_row = make_bold_names(sig_norm_counts_croI_annotation, rownames, bold_genes$rowname),
         labels_col = labels_cols$Strain,
         scale = "row") # Plots Z-scores rather than actual normalized count values

```

-----------------------------------------

### 8. Session information

```{r Session info}
sessionInfo()
```

-----------------------------------------

### 9. References 

```{r References}
packages <- c("tidyverse", "DESeq2", "pheatmap", "ggsci", "IHW", "apeglm", "ggrepel", "WGCNA")

packages %>%
  map(citation) %>%
  print(style = "text")
```