---
title: "RNA-seq analysis - Pathway enrichment analysis"
author: "Jorge Peña Díaz"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

-----------------------------------------

Analysis used to compare the gene expression profile of the ∆*croI* and wild-type (WT) *C. rodentium* DB1000 strains as published in:

> Peña‑Díaz, J., Woodward, S.E., Creus‑Cuadros, A., Serapio‑Palacios, A., Deng, W., and Finlay, B.B. (2023). Quorum Sensing Modulates
Bacterial Virulence and Colonization Dynamics During an Enteric Infection. bioRxiv, [10.1101/2023.03.14.532656](https://www.biorxiv.org/content/10.1101/2023.03.14.532656v1).


-----------------------------------------

### 1. Setup

Load the required packages for the analysis.

```{r eval = TRUE, message=FALSE, warning=FALSE}
library(tidyverse)
library(EnrichmentBrowser)
library(clusterProfiler)
library(ggsci)
```


### 2. Import the required data

```{r Raw data import, warning=FALSE}
# Import differentially regulated genes
croI_res_sig <- read.delim("data/DEGs_dcroI.txt")

# Import gene expression
croI_res_all <- read.delim("data/ExpressionData_dcroI.txt")
```


### 3. Gene Set Enrichment Analysis (GSEA)

#### 3.1 Data preparation
Create the required files before analysis

```{r Ranked qn}

# Create a ranked list of genes arranged according to the Wald-test statistic
croI_res_all %>%
  select(old_locus_tag, stat) %>%
  na.omit() %>% 
  filter(is.finite(stat)) %>%
  arrange(stat)  %>%
  write.table("data/croI.rnk",sep="\t",col.names = FALSE, row.names=FALSE,quote=FALSE)


# Create a gene set matrix (GMT) by using KEGG
cro <- getGenesets(org = "cro", db = "kegg", cache = TRUE)

# Write GMT file
writeGMT(cro, gmt.file = "data/Citrobacter_rodentium.gmt")
```

#### 3.2 GSEA 
Run the gene set enrichment analysis in the terminal

```{}
GSEA_4.2.3/gsea-cli.sh GSEAPreRanked \
-gmx data/Citrobacter_rodentium.gmt \
-rnk data/croI.rnk \
-collapse false \
-nperm 100000 \
-scoring_scheme weighted \
-rpt_label croI \
-plot_top_x 20 \
-rnd_seed 12345 \
-set_max 500 \
-set_min 5 \
-zip_report false \
-out croI_GSEA > gsea_output.txt
```

#### 3.3 Import results for data vizualization 

```{r GSEA Results}
# Import results
Upregulated <- read.delim("data/GSEA/gsea_report_for_na_pos.tsv") %>%
  mutate(Direction = "Upregulated")

Downregulated <- read.delim("data/GSEA/gsea_report_for_na_neg.tsv") %>%
   mutate(Direction = "Downregulated")

# Clean data
GSEA <- bind_rows(Upregulated, Downregulated) %>%
  mutate(Pathway = str_sub(NAME, end = 8)) %>%
  mutate(NAME = str_sub(NAME, start = 10)) %>%
  mutate(NAME = str_replace_all(NAME, "_", " ")) %>%
  mutate(NAME = str_to_sentence(NAME)) %>%
  select(NAME, Pathway, SIZE, ES, NES, NOM.p.val, q.val = "FDR.q.val", p.adj = "FWER.p.val",
         RANK.AT.MAX, LEADING.EDGE, Direction) %>%
  filter(q.val <= 0.05) 
```

### 4. Over-representation analysis (ORA)

#### 4.1 Data preparation
Create the required files before running the over-representation analysis 

```{r over-representation analysis}
#  List of all background genes
universe <- croI_res_all %>%
  as.data.frame() %>%
  select(old_locus_tag, stat) %>%
  na.omit() %>% 
  filter(is.finite(stat)) %>%
  arrange(desc(stat))

# Format the data
universe_geneList = universe[,2]
names(universe_geneList) = as.character(universe[,1])
universe_geneList <- names(sort(universe_geneList, decreasing = TRUE))

# List of differentially regulated genes
d <- croI_res_sig %>%
  select(old_locus_tag, stat) %>%
  na.omit() %>% 
  filter(is.finite(stat)) %>%
  arrange(desc(stat)) %>%
  as.data.frame()

# Format the data
geneList = d[,2]
names(geneList) = as.character(d[,1])

# Subset up- and down-regulated genes
geneList_Up <- geneList[geneList > 0]
geneList_Down <- geneList[geneList < 0]

# Sort in decreasing order
geneList_Up <- names(sort(abs(geneList_Up), decreasing = TRUE))
geneList_Down <- names(sort(abs(geneList_Down), decreasing = TRUE))
```

#### 3.2 ORA (KEGG pathways)

KEGG pathway over-representation analysis

```{r ORA_KEGG pathways}
KEGG_up <- enrichKEGG(gene = geneList_Up,
                      organism = 'cro',
                      minGSSize = 5,
                      universe = universe_geneList,
                      pvalueCutoff = 0.05)

KEGG_down <- enrichKEGG(gene = geneList_Down,
                        organism = 'cro',
                        minGSSize = 5,
                        universe = universe_geneList,
                        pvalueCutoff = 0.05)

# Merge data
KEGG_up <- KEGG_up %>%
  mutate(Direction = "Upregulated") %>%
  as.data.frame()

KEGG_down <- KEGG_down %>%
  mutate(Direction = "Downregulated") %>%
  as.data.frame()

KEGG <- bind_rows(KEGG_up, KEGG_down)
``` 

#### 3.2 ORA (KEGG modules)

KEGG module over-representation analysis

```{r ORA_KEGG modules}
# KEGG Module over-representation
Module_up <- enrichMKEGG(gene = geneList_Up,
                         organism = 'cro',
                         universe = universe_geneList,
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.05)

Module_down <- enrichMKEGG(gene = geneList_Down,
                           organism = 'cro',
                           universe = universe_geneList,
                           pAdjustMethod = "BH",
                           pvalueCutoff = 0.05)

# Merge data
Module_up <- Module_up %>%
  mutate(Direction = "Upregulated") %>%
  as.data.frame()

Module_down <- Module_down %>%
  mutate(Direction = "Downregulated") %>%
  as.data.frame()

Module <- bind_rows(Module_up, Module_down)
```


### 5. Data visualization 

We will combine the results from GSEA and module over-representation analysis and then plot them in a lollipop plot.

```{r GSEA_Module data, fig.dim = c(9.5, 6)}
# Clean GSEA data
GSEA_clean <- GSEA %>%
  mutate(p.adj = ifelse(p.adj == 0, 0.00001, p.adj)) %>%  # Replace since we cannot calculate actual p value
  mutate(q.val = ifelse(q.val == 0, 0.0000000001, q.val)) %>% # Replace since we cannot calculate actual q value
  select(NAME, Pathway, p.adj, q.val, Direction, SIZE) %>%
  mutate(Analysis = "GSEA")

# Clean KEGG Module over-representation data
Module_clean <- Module %>%
  rename(NAME = "Description", Pathway = "ID", p.adj = "p.adjust", q.val = "qvalue") %>%
  select(NAME, Pathway, p.adj, q.val, Direction, Count) %>%
  rename(SIZE = "Count") %>%
  mutate(Analysis = "Module")

# Merge the data and do final cleaning before plotting
Data_clean <- rbind(GSEA_clean, Module_clean) %>%
  mutate(NAME = str_replace_all(NAME, " ", "\n")) %>%
  mutate(NAME = str_replace(NAME, "EHEC/EPEC\npathogenicity\nsignature,\nT3SS\nand\neffectors", 
                                   "EHEC/EPEC\npathogenicity signature\n(T3SS and effectors)")) %>%
  mutate(NAME = str_replace(NAME, "Starch\nand\nsucrose\nmetabolism",
                            "Starch and\nsucrose\nmetabolism")) %>%
  mutate(qscore = -log10(p.adj)) %>% 
  mutate(Direction = factor(Direction, levels = c("Upregulated", "Downregulated"))) 
  
# Plot the data
Data_clean %>%
  ggplot(aes(x = qscore, reorder(NAME, q.val))) +
  geom_linerange(aes(xmin = 0, xmax = qscore, colour = Analysis, y = reorder(NAME, q.val, decreasing = TRUE)), linewidth = 2.5) +
  geom_point(aes(size = SIZE, colour = Analysis)) +
  xlab("-log10(p-value)") +
  ylab("") +
  facet_grid(Direction~., scales = "free", space = "free") +
  labs(colour = "Analysis") +
  theme_bw() +
  scale_y_discrete() +
  scale_colour_d3(labels = c("GSEA (KEGG Pathways)", "ORA (KEGG Modules)")) +
  scale_x_continuous(limits = c(0,25)) +
  scale_size(range = c(1.5, 9.5), name="Size", limits = c(5, 60)) +
  theme(plot.title = element_text(size = rel(2), hjust = 0.5),
        axis.title = element_text(size = rel(1.5)),
        axis.text.x = element_text(size = rel(1.8)),
        axis.text.y = element_text(size = rel(1.8), hjust = 1),
        legend.title = element_text(size = rel(1.5)),
        legend.text = element_text(size = rel(1.15)),
        strip.text = element_text(size = rel(1.125)))  +
  guides(colour = guide_legend(order = 1))
```

-----------------------------------------

### 6. Session information

```{r Session info}
sessionInfo()
```

-----------------------------------------

### 7. References 

```{r References}
packages <- c("tidyverse", "EnrichmentBrowser", "clusterProfiler", "ggsci")

packages %>%
  map(citation) %>%
  print(style = "text")